src/main/java/uk/gov/hmcts/reform/divorce/orchestration/service/impl/CaseOrchestrationServiceImpl.java

```
@@ -773,24 +776,35 @@ public CcdCallbackResponse confirmSolDnReviewPetition(CaseDetails caseDetails) {
-                if (featureToggleService.isFeatureEnabled(Features.OBJECT_TO_COSTS)
-                    && hasJudgeMadeCostsDecision(caseDetails.getCaseData())) {
-                    caseData.putAll(documentGenerationWorkflow.run(
-                        caseDetails, authToken, COSTS_ORDER_DOCUMENT_TYPE, COSTS_ORDER_JUDGE, COSTS_ORDER_DOCUMENT_TYPE));
-                } else {
-                    caseData.putAll(documentGenerationWorkflow.run(
-                        caseDetails, authToken, COSTS_ORDER_DOCUMENT_TYPE, COSTS_ORDER, COSTS_ORDER_DOCUMENT_TYPE));
-                }
+            caseData.putAll(documentGenerationWorkflow.run(caseDetails, authToken, DECREE_NISI_DOCUMENT_TYPE, DECREE_NISI, DECREE_NISI_FILENAME));
+
+            if (shouldCostOrderBeSent(caseData)) {
+                sendCostOrderDocument(authToken, caseDetails, caseData);
             }
         }
 
         return caseData;
     }
 
+    private void sendCostOrderDocument(String authToken, CaseDetails caseDetails, Map<String, Object> caseData) throws WorkflowException {
+        if (featureToggleService.isFeatureEnabled(Features.OBJECT_TO_COSTS)
+            && hasJudgeGrantedCostsDecision(caseDetails.getCaseData())) {
+            caseData.putAll(documentGenerationWorkflow.run(
+                caseDetails, authToken, COSTS_ORDER_DOCUMENT_TYPE, COSTS_ORDER_JUDGE, COSTS_ORDER_DOCUMENT_TYPE));
+        } else {
+            caseData.putAll(documentGenerationWorkflow.run(
+                caseDetails, authToken, COSTS_ORDER_DOCUMENT_TYPE, COSTS_ORDER, COSTS_ORDER_DOCUMENT_TYPE));
+        }
+    }
+
+    private boolean shouldCostOrderBeSent(Map<String, Object> caseData) {
+        if (featureToggleService.isFeatureEnabled(Features.OBJECT_TO_COSTS)
+            && (isJudgeCostClaimEmpty(caseData) && !isCostClaimGrantedPopulated(caseData)) || isJudgeCostClaimAdjourned(caseData)) {
+            return false;
+        }
+        return isPetitionerClaimingCosts(caseData);
+    }
+
```

@yarpo I have update the logic to make sure nothing gets through without a feature toggle.

Previously, i relied on the implicit toggling that CCD would do as the fields would not be available until the release. This would mean it was always return false and all back to the current functionality. 

However, for peace of mind i have added the toggle to the `shouldCostOrderBeSent` check as well

---

Suggestion: I'd rather fail fast to reduce the complexity of the logic here. Also if Sonar becomes aggressive with condition checks, might want to try something like this (or a variation)

`private boolean shouldCostOrderBeSent(Map<String, Object> caseData) {
          if (featureToggleService.isFeatureEnabled(Features.OBJECT_TO_COSTS)){
           return false;
          }

          if(isJudgeCostClaimEmpty(caseData) 
             && !isCostClaimGrantedPopulated(caseData)) 
             || isJudgeCostClaimAdjourned(caseData)) {
              return false;
           } else {
             return isPetitionerClaimingCosts(caseData);
           }   
       }`

---

I think it would make sense to test with OtC feature toggle on and off.

---

for off - behaviour shouldn't change

---

The unit tests have the feature toggle included. They still pass in past expected scenarios without the inclusion of the feature toggle. Which to me shows its no behaviour change nature